File: wqr/mediation.py
"""
Wavelet Quantile Mediation & Moderation Regression.

Implements:
  - Direct effect:    Y ~ X
  - Interaction:      Y ~ X + Z + X·Z  (moderation)
  - Path a:           Z ~ X
  - Path b:           Y ~ X + Z        (coef on Z)
  - Indirect effect:  a × b            (mediation)

Based on Adebayo (2025):
  "Can ESG Uncertainty Alter the Emissions Impact of Renewable Energy
   Consumption?"  Statistical Journal of the IAOS.
  doi:10.1177/18747655261445375
"""

import numpy as np
import pandas as pd
from dataclasses import dataclass
from typing import Optional

import statsmodels.api as sm
from statsmodels.regression.quantile_regression import QuantReg

from .utils import modwt_mra, aggregate_bands, safe_print


@dataclass
class MediationResult:
    """Container for wavelet quantile mediation + moderation results."""
    direct: pd.DataFrame         # Y ~ X
    interaction: pd.DataFrame    # coef on X·Z in Y ~ X + Z + XZ
    path_a: pd.DataFrame         # Z ~ X
    path_b: pd.DataFrame         # coef on Z in Y ~ X + Z
    indirect: pd.DataFrame       # a × b with joint p-value
    quantiles: np.ndarray
    bands: list
    dep_name: str
    main_name: str
    mod_name: str
    n_obs: int

    def summary(self):
        """Print summary."""
        lines = [
            "",
            "╔══════════════════════════════════════════════════════════╗",
            "║  Wavelet Quantile Mediation & Moderation — Summary      ║",
            "╚══════════════════════════════════════════════════════════╝",
            "",
            f"  Dependent (Y) : {self.dep_name}",
            f"  Main (X)      : {self.main_name}",
            f"  Moderator (Z) : {self.mod_name}",
            f"  Observations  : {self.n_obs}",
            f"  Bands         : {self.bands}",
            "",
        ]
        for name, df in [("Direct", self.direct), ("Interaction", self.interaction),
                         ("Path a", self.path_a), ("Path b", self.path_b),
                         ("Indirect", self.indirect)]:
            n_sig = (df["p_value"] < 0.05).sum() if "p_value" in df.columns else 0
            lines.append(f"  {name:14s}  sig(5%)={n_sig}/{len(df)}")
        lines.append("")
        safe_print("\n".join(lines))
        return self


def _qr_coef(dep, indep_dict, tau, n_boot=100):
    """Run QR and return dict {var_name: (beta, pval)} for each regressor."""
    X_cols = list(indep_dict.keys())
    X_mat = np.column_stack([indep_dict[k] for k in X_cols])
    X_mat = sm.add_constant(X_mat)

    try:
        model = QuantReg(dep, X_mat)
        res = model.fit(q=tau, max_iter=1000, p_tol=1e-5)
        out = {}
        for i, name in enumerate(X_cols):
            idx = i + 1
            out[name] = (
                res.params[idx],
                res.bse[idx] if res.bse is not None else np.nan,
                res.pvalues[idx] if res.pvalues is not None else np.nan,
            )
        return out
    except Exception:
        return {k: (np.nan, np.nan, np.nan) for k in X_cols}


def wavelet_mediation(y, x, z, quantiles=None, wavelet="la8", J=None,
                      n_boot=100, dep_name="Y", main_name="X",
                      mod_name="Z", verbose=True):
    """
    Wavelet Quantile Mediation & Moderation Regression.

    Parameters
    ----------
    y : array-like
        Dependent variable (Y).
    x : array-like
        Main independent variable (X).
    z : array-like
        Mediator / Moderator variable (Z).
    quantiles : array-like or None
        Quantile grid.
    wavelet : str
    J : int or None
    n_boot : int
    dep_name, main_name, mod_name : str
        Variable names for labels.
    verbose : bool

    Returns
    -------
    MediationResult
    """
    y = np.asarray(y, dtype=np.float64)
    x = np.asarray(x, dtype=np.float64)
    z = np.asarray(z, dtype=np.float64)
    n_obs = len(y)

    if J is None:
        J = max(3, int(np.log2(n_obs)) - 3)

    if quantiles is None:
        quantiles = np.array([0.01, 0.05, 0.10, 0.20, 0.30, 0.40, 0.50,
                              0.60, 0.70, 0.80, 0.90, 0.95, 0.99])

    if verbose:
        print(f"Wavelet Quantile Mediation/Moderation  (wf={wavelet}, J={J})")
        print(f"  {main_name} → {mod_name} → {dep_name}  (Mediation)")
        print(f"  {main_name}×{mod_name} → {dep_name}  (Moderation)")

    # Decompose
    det_y, _ = modwt_mra(y, wavelet=wavelet, level=J)
    det_x, _ = modwt_mra(x, wavelet=wavelet, level=J)
    det_z, _ = modwt_mra(z, wavelet=wavelet, level=J)

    bands_y = aggregate_bands(det_y)
    bands_x = aggregate_bands(det_x)
    bands_z = aggregate_bands(det_z)
    band_names = list(bands_y.keys())

    rec_direct, rec_int, rec_a, rec_b, rec_ind = [], [], [], [], []

    for band in band_names:
        yb = bands_y[band]
        xb = bands_x[band]
        zb = bands_z[band]
        xzb = xb * zb

        for tau in quantiles:
            # Direct: Y ~ X
            r = _qr_coef(yb, {"x": xb}, tau, n_boot)
            rec_direct.append({"quantile": tau, "band": band,
                               "beta": r["x"][0], "se": r["x"][1], "p_value": r["x"][2]})

            # Interaction: Y ~ X + Z + XZ  (report XZ coef)
            r = _qr_coef(yb, {"x": xb, "z": zb, "xz": xzb}, tau, n_boot)
            rec_int.append({"quantile": tau, "band": band,
                            "beta": r["xz"][0], "se": r["xz"][1], "p_value": r["xz"][2]})

            # Path a: Z ~ X
            r = _qr_coef(zb, {"x": xb}, tau, n_boot)
            a_beta, a_se, a_pval = r["x"]
            rec_a.append({"quantile": tau, "band": band,
                          "beta": a_beta, "se": a_se, "p_value": a_pval})

            # Path b: Y ~ X + Z (report Z coef)
            r = _qr_coef(yb, {"x": xb, "z": zb}, tau, n_boot)
            b_beta, b_se, b_pval = r["z"]
            rec_b.append({"quantile": tau, "band": band,
                          "beta": b_beta, "se": b_se, "p_value": b_pval})

            # Indirect: a × b
            indirect_beta = a_beta * b_beta
            joint_p = max(a_pval, b_pval) if not (np.isnan(a_pval) or np.isnan(b_pval)) else np.nan
            rec_ind.append({"quantile": tau, "band": band,
                            "beta": indirect_beta, "p_value": joint_p})

        if verbose:
            print(f"  ✓ Band {band} done")

    if verbose:
        print("  ✓ Mediation/Moderation completed.")

    return MediationResult(
        direct=pd.DataFrame(rec_direct),
        interaction=pd.DataFrame(rec_int),
        path_a=pd.DataFrame(rec_a),
        path_b=pd.DataFrame(rec_b),
        indirect=pd.DataFrame(rec_ind),
        quantiles=quantiles,
        bands=band_names,
        dep_name=dep_name,
        main_name=main_name,
        mod_name=mod_name,
        n_obs=n_obs,
    )

File: R/mediation.R
#' @title Wavelet Quantile Mediation and Moderation
#'
#' @description Fits five band-by-quantile quantile regressions on a
#'   triplet \eqn{(y, x, z)}:
#'   \itemize{
#'     \item \strong{Direct}: \eqn{y = \alpha + \beta_{d} x}
#'     \item \strong{Interaction (moderation)}:
#'           \eqn{y = \alpha + b_x x + b_z z + b_{xz} (x z)}
#'     \item \strong{Path a}: \eqn{z = \alpha + a x}
#'     \item \strong{Path b}: \eqn{y = \alpha + b_x x + b_{z} z}
#'     \item \strong{Indirect}: \eqn{a \cdot b_z}
#'   }
#'   Each regression is fit on MODWT-aggregated bands.
#'
#' @param y,x,z Numeric vectors of equal length.
#' @param quantiles Numeric vector of \eqn{\tau} in (0, 1).
#' @param wavelet Character.
#' @param J Integer or \code{NULL} (auto-selected from sample size).
#' @param dep_name,main_name,mod_name Character labels.
#' @param verbose Logical.
#'
#' @return Object of class \code{"mediation_result"}.
#'
#' @references Adebayo, T.S. (2025).
#'   \emph{Statistical Journal of the IAOS}.
#'   \doi{10.1177/18747655261445375}.
#'
#' @examples
#' set.seed(1); n <- 200
#' x <- rnorm(n); z <- 0.4 * x + rnorm(n, sd = 0.5)
#' y <- 0.3 * x + 0.4 * z + 0.2 * x * z + rnorm(n, sd = 0.5)
#' fit <- wavelet_mediation(y, x, z,
#'                          quantiles = c(0.25, 0.5, 0.75),
#'                          wavelet = "la8", J = 4, verbose = FALSE)
#' print(fit)
#'
#' @export
#' @importFrom quantreg rq
wavelet_mediation <- function(y, x, z,
                              quantiles = c(0.05, 0.10, 0.25, 0.50,
                                            0.75, 0.90, 0.95),
                              wavelet = "la8", J = NULL,
                              dep_name = "Y", main_name = "X",
                              mod_name = "Z", verbose = TRUE) {
  y <- as.numeric(y); x <- as.numeric(x); z <- as.numeric(z)
  n_obs <- length(y)
  if (is.null(J)) J <- max(3L, as.integer(floor(log2(n_obs)) - 3L))
  if (verbose)
    message("Wavelet Mediation/Moderation (wf=", wavelet, ", J=", J, ")")

  dec_y <- modwt_mra(y, wavelet, J)
  dec_x <- modwt_mra(x, wavelet, J)
  dec_z <- modwt_mra(z, wavelet, J)
  by <- aggregate_bands(dec_y$details)
  bx <- aggregate_bands(dec_x$details)
  bz <- aggregate_bands(dec_z$details)
  band_names <- names(by)

  rec_d <- rec_i <- rec_a <- rec_b <- rec_ind <- list()
  qr_one <- function(yv, design, tau, which_col) {
    fit <- tryCatch(quantreg::rq(yv ~ design, tau = tau, method = "br"),
                    error = function(e) NULL)
    if (is.null(fit)) return(c(NA_real_, NA_real_, NA_real_))
    s <- tryCatch(summary(fit, se = "iid"), error = function(e) NULL)
    if (is.null(s) || nrow(s$coefficients) < which_col + 1L)
      return(c(NA_real_, NA_real_, NA_real_))
    cf <- s$coefficients
    c(cf[which_col + 1L, 1], cf[which_col + 1L, 2], cf[which_col + 1L, 4])
  }

  for (b in band_names) {
    yb <- by[[b]]; xb <- bx[[b]]; zb <- bz[[b]]
    xzb <- xb * zb
    for (tau in quantiles) {
      d  <- qr_one(yb, xb, tau, 1L)
      i  <- qr_one(yb, cbind(xb, zb, xzb), tau, 3L)
      a  <- qr_one(zb, xb, tau, 1L)
      bp <- qr_one(yb, cbind(xb, zb), tau, 2L)
      rec_d  [[length(rec_d) + 1]]  <- list(quantile = tau, band = b,
        beta = d[1], se = d[2], p_value = d[3])
      rec_i  [[length(rec_i) + 1]]  <- list(quantile = tau, band = b,
        beta = i[1], se = i[2], p_value = i[3])
      rec_a  [[length(rec_a) + 1]]  <- list(quantile = tau, band = b,
        beta = a[1], se = a[2], p_value = a[3])
      rec_b  [[length(rec_b) + 1]]  <- list(quantile = tau, band = b,
        beta = bp[1], se = bp[2], p_value = bp[3])
      ind_beta <- a[1] * bp[1]
      ind_p <- if (any(!is.finite(c(a[3], bp[3])))) NA_real_
        else max(a[3], bp[3])
      rec_ind[[length(rec_ind) + 1]] <- list(quantile = tau, band = b,
        beta = ind_beta, p_value = ind_p)
    }
    if (verbose) message("  band ", b, " done")
  }

  to_df <- function(L) do.call(rbind,
                               lapply(L, as.data.frame,
                                      stringsAsFactors = FALSE))
  res <- list(direct = to_df(rec_d),
              interaction = to_df(rec_i),
              path_a = to_df(rec_a),
              path_b = to_df(rec_b),
              indirect = to_df(rec_ind),
              quantiles = quantiles, bands = band_names,
              dep_name = dep_name, main_name = main_name,
              mod_name = mod_name, n_obs = n_obs,
              wavelet = wavelet, J = J,
              call = match.call(),
              method = "Wavelet Quantile Mediation & Moderation")
  class(res) <- "mediation_result"
  res
}


#' @export
print.mediation_result <- function(x, ...) {
  cat("\nWavelet Quantile Mediation & Moderation\n")
  cat(strrep("=", 48), "\n", sep = "")
  cat("  Y =", x$dep_name, "  X =", x$main_name, "  Z =", x$mod_name, "\n")
  cat("  wavelet =", x$wavelet, "  J =", x$J,
      "  N =", x$n_obs, "  q-grid =", length(x$quantiles),
      "  bands =", paste(x$bands, collapse = ","), "\n")
  for (nm in c("direct", "interaction", "path_a", "path_b", "indirect")) {
    df <- x[[nm]]; n_sig <- sum(df$p_value < 0.05, na.rm = TRUE)
    cat(sprintf("  %-12s  sig(5%%) = %d / %d\n",
                nm, n_sig, sum(is.finite(df$p_value))))
  }
  invisible(x)
}

#' @export
summary.mediation_result <- function(object, ...) { print(object); invisible(object) }

#' @export
plot.mediation_result <- function(x, ...) plot_mediation_panel(x, ...)
